src/analyze/analyze-chid.c

```diff
@@ -0,0 +1,136 @@
+                [CHID_SMBIOS_FAMILY]                 = "product_family",
+                [CHID_SMBIOS_PRODUCT_NAME]           = "product_name",
+                [CHID_SMBIOS_PRODUCT_SKU]            = "product_sku",
+                [CHID_SMBIOS_BASEBOARD_MANUFACTURER] = "board_vendor",
+                [CHID_SMBIOS_BASEBOARD_PRODUCT]      = "board_name",
+        };
+
+        _cleanup_(table_unrefp) Table *table = NULL;
+        int r;
+
+        if (detect_container() > 0)
+                return log_error_errno(SYNTHETIC_ERRNO(ENOTRECOVERABLE), "Container environments do not have SMBIOS.");
+
+        table = table_new("type", "chid");
+        if (!table)
+                return log_oom();
+
+        (void) table_set_align_percent(table, table_get_cell(table, 0, 0), 100);
+        (void) table_set_align_percent(table, table_get_cell(table, 0, 1), 50);
+
+        _cleanup_close_ int smbios_fd = open("/sys/class/dmi/id", O_RDONLY|O_DIRECTORY|O_CLOEXEC);
+        if (smbios_fd < 0)
+                return log_error_errno(errno, "Failed to open SMBIOS sysfs object: %m");
+
+        _cleanup_(smbios_fields_free) char16_t* smbios_fields[_CHID_SMBIOS_FIELDS_MAX] = {};
+        for (ChidSmbiosFields f = 0; f < _CHID_SMBIOS_FIELDS_MAX; f++) {
+                _cleanup_free_ char *buf = NULL;
+                size_t size;
+
+                r = read_virtual_file_at(smbios_fd, smbios_files[f], SIZE_MAX, &buf, &size);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to read SMBIOS field '%s': %m", smbios_files[f]);
+
+                if (size < 1 || buf[size-1] != '\n')
+                        return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "Expected SMBIOS field '%s' to end in newline, but it doesn't, refusing.", smbios_files[f]);
+
+                size--;
+
+                smbios_fields[f] = utf8_to_utf16(buf, size);
+                if (!smbios_fields[f])
+                        return log_oom();
+        }
+
+        EFI_GUID chids[CHID_TYPES_MAX] = {};
+        chid_calculate((const char16_t* const*) smbios_fields, chids);
+
+        if (strv_isempty(strv_skip(argv, 1)))
+                for (size_t t = 0; t < CHID_TYPES_MAX; t++) {
+                        r = add_chid(table, chids, t);
+                        if (r < 0)
+                                return r;
+                }
+        else {
+                STRV_FOREACH(as, strv_skip(argv, 1)) {
+                        size_t t;
+                        r = parse_chid_type(*as, &t);
+                        if (r < 0)
+                                return log_error_errno(r, "Failed to pare CHID type: %s", *as);
+
+                        r = add_chid(table, chids, t);
+                        if (r < 0)
+                                return r;
+                }
+
+                (void) table_set_sort(table, (size_t) 0);
+        }
+
+        r = table_print_with_pager(table, arg_json_format_flags, arg_pager_flags, arg_legend);
```

Comment:
I am not sure it should. You cannot really use this output as is anyway, we wouldn't know what to fill in in the compatible string or the name fields, and after all you typically only want to pick a single uuid from the full list. moreover when looking at this data you want to know which uuid is which chid type, but for the ukify stuff this doesn't matter really.

hence, i sense a better approach might be to include a "jq" based cmdline in the man page maybe that shows how to select one of the chids in this output, then augment it with a compatible string and a name to make it suitable for ukify. 

i mean, yes, this is related data, but only some of the fields, the document structure should be quite different i am sure.

Reply:
> we wouldn't know what to fill in in the compatible string or the name fields

Compatible can be extracted from `cat /sys/firmware/devicetree/base/compatible | strings | head -1` or something.

> and after all you typically only want to pick a single uuid from the full list

Actually, no. In most cases (due to different SKUs which are still compatible from the kernel/devicetree point of view) more than one is usefule.

> moreover when looking at this data you want to know which uuid is which chid type,

If only json supported comments... `"type" : 3` isn't a very good explanation IMHO. fwupdtool, for example, prints them as following:
```
{0ca52c05-0cb3-5309-96a6-86e9c5c886c0}   <- Manufacturer + Family + ProductName + ProductSku + BaseboardManufacturer + BaseboardProduct
{57582890-0a06-5dd8-9583-badac8a9f2eb}   <- Manufacturer + Family + ProductName + ProductSku
{66e29b54-1fc9-5453-b311-6727cdec8ded}   <- Manufacturer + Family + ProductName
{2a73f48e-8b58-5ba3-bf84-3f5ced9942c3}   <- Manufacturer + ProductSku + BaseboardManufacturer + BaseboardProduct
{36d8d875-a363-520a-82f1-c2ced7f75105}   <- Manufacturer + ProductSku
{f620a38a-b5f1-50fb-ab3c-35b15d2e63ea}   <- Manufacturer + ProductName + BaseboardManufacturer + BaseboardProduct
{d24c81d8-b6c8-5f40-842d-dbb8b206df84}   <- Manufacturer + ProductName
{2a73f48e-8b58-5ba3-bf84-3f5ced9942c3}   <- Manufacturer + Family + BaseboardManufacturer + BaseboardProduct
{36d8d875-a363-520a-82f1-c2ced7f75105}   <- Manufacturer + Family
{60ecfe16-87a0-52aa-84b2-dead2df99592}   <- Manufacturer + EnclosureKind
{7f960b9f-0f9d-51a7-8d42-6118dedc84ef}   <- Manufacturer + BaseboardManufacturer + BaseboardProduct
{708fe7d4-e5a2-579b-9a0e-71cb222d51db}   <- Manufacturer
Extra Hardware IDs
------------------
{b9c1762c-7dea-5fe1-a11c-99ba308e6ef2}   <- Manufacturer + Family + ProductName + ProductSku + BiosVendor
{3aa7c106-e5d6-5ae9-aed5-708dcbb3e103}   <- Manufacturer + Family + ProductName + BiosVendor
{259c41f9-6b6b-56fe-a853-39fbe7acbd69}   <- Manufacturer + BiosVendor
```

Reply:
> Compatible can be extracted from `cat /sys/firmware/devicetree/base/compatible | strings | head -1` or something.

On DT, but the CHID analysis tool is generic and supposed to be independent of the DT usecase.

> Actually, no. In most cases (due to different SKUs which are still compatible from the kernel/devicetree point of view) more than one is useful

More then one: yes. All? definitely not. because if you use the manufacturer one then you might as well not bother with the others. So this *definitely* needs some careful selection by somebody putting together the matches, otherwise it would be useless.

Reply:
> If only json supported comments... `"type" : 3` isn't a very good explanation IMHO. fwupdtool, for example, prints them as following:

we can add another field to encode this.

Reply:
I extended the table now to visually indicate which smbios fields go into which chid.